File: python/kohaku/src/kohaku/llm/retry.py
"""Jitter-added exponential backoff retry limited to transient PROVIDER-side failures (port of TS adapters/retry.ts).

This single file centralizes the retry policy. Design-level separations:
- ABORTED (caller cancellation / timeout exceeded) propagates immediately and is not retried.
- INVALID_OUTPUT / CONFIG are out of scope (schema-derived failures are not fixed by retrying).
- `is_retryable=False` PROVIDER (e.g. ollama's structured-output 400) is also out of scope. That is a
  structured-output incompatibility handled by the prompt JSON fallback path (auto). Retrying here would
  cause a double spend of "fallback × backoff", so it propagates immediately.
- The whole sequence never exceeds the timeout (deadline). If the next wait would not fit the remaining budget, give up.
"""

from __future__ import annotations

import math
import random as _random
import time
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from email.utils import parsedate_to_datetime

from .abort import AbortError, AbortSignal, default_sleep
from .env import RetryPolicy


class ApiCallError(Exception):
    """A failure of an OpenAI-compatible API call (equivalent to the AI SDK's APICallError).

    Always carries `is_retryable` (bool), along with `status_code` / `response_headers`.
    Retryability and Retry-After are decided from this information.
    """

    def __init__(
        self,
        message: str,
        *,
        is_retryable: bool,
        status_code: int | None = None,
        response_headers: dict[str, str] | None = None,
        url: str | None = None,
    ) -> None:
        super().__init__(message)
        self.is_retryable = is_retryable
        self.status_code = status_code
        self.response_headers: dict[str, str] = response_headers if response_headers is not None else {}
        self.url = url


type SleepFn = Callable[[float, AbortSignal], Awaitable[None]]


def _monotonic_ms() -> float:
    return time.monotonic() * 1000.0


@dataclass(frozen=True)
class RetryDeps:
    """Injectable side effects for determinism (tests substitute the wait, clock, and randomness)."""

    sleep: SleepFn = default_sleep
    now: Callable[[], float] = _monotonic_ms
    random: Callable[[], float] = _random.random


default_retry_deps = RetryDeps()


def _round_half_up(x: float) -> int:
    """Match JS's Math.round (rounds halves toward +∞)."""
    return math.floor(x + 0.5)


def _is_abort_like(err: BaseException, signal: AbortSignal) -> bool:
    """Whether the error originates from an abort/timeout (decided by signal being aborted or by AbortError)."""
    if signal.aborted:
        return True
    return isinstance(err, AbortError)


def _find_api_call_error(err: object, depth: int = 0) -> object | None:
    """Structurally search for an ApiCallError (also following the cause chain)."""
    if err is None or depth > 4:
        return None
    if isinstance(err, ApiCallError):
        return err
    is_retryable = getattr(err, "is_retryable", None)
    if isinstance(is_retryable, bool) and (
        hasattr(err, "response_headers") or hasattr(err, "status_code") or hasattr(err, "url")
    ):
        return err
    return _find_api_call_error(getattr(err, "__cause__", None), depth + 1)


def _retry_after_from_headers(headers: dict[str, str] | None) -> float | None:
    """Interpret Retry-After-like headers (retry-after-ms / retry-after) as milliseconds."""
    if not headers:
        return None
    retry_after_ms = headers.get("retry-after-ms")
    if retry_after_ms is not None:
        try:
            ms = float(retry_after_ms)
        except ValueError:
            ms = float("nan")
        if math.isfinite(ms) and ms >= 0:
            return ms
    retry_after = headers.get("retry-after")
    if retry_after is not None:
        # Seconds (e.g. "2") or an HTTP date (e.g. "Wed, 21 Oct 2025 07:28:00 GMT").
        try:
            seconds = float(retry_after)
        except ValueError:
            seconds = float("nan")
        if math.isfinite(seconds) and seconds >= 0:
            return seconds * 1000.0
        try:
            parsed = parsedate_to_datetime(retry_after)
            date_ms = parsed.timestamp() * 1000.0 - time.time() * 1000.0
        except (TypeError, ValueError):
            return None
        if math.isfinite(date_ms) and date_ms >= 0:
            return date_ms
    return None


def retryable_provider_error(err: object) -> tuple[bool, float | None]:
    """Decide whether this is a retryable PROVIDER failure.

    Return value (retryable, retry_after_ms):
    - (False, None): not retryable (non-ApiCallError / is_retryable=False)
    - (True, None): retryable, no Retry-After (leave it to exponential backoff)
    - (True, number): retryable, with Retry-After (respect this wait)
    """
    api = _find_api_call_error(err)
    if api is None or getattr(api, "is_retryable", None) is not True:
        return (False, None)
    return (True, _retry_after_from_headers(getattr(api, "response_headers", None)))


def is_retryable_provider_error(err: object) -> bool:
    """Whether this is a retryable PROVIDER failure (the boolean version used by the auto fallback branch)."""
    return retryable_provider_error(err)[0]


def next_delay_ms(
    policy: RetryPolicy, attempt: int, retry_after_ms: float | None, rand: Callable[[], float]
) -> int:
    """The next backoff wait (milliseconds). If Retry-After is present, respect it; otherwise perturb
    base = initial_delay_ms * backoff_factor^attempt by ±jitter."""
    if retry_after_ms is not None:
        return max(0, _round_half_up(retry_after_ms))
    base = policy.initial_delay_ms * policy.backoff_factor**attempt
    # Map rand() ∈ [0,1) to [-1,1) and perturb as a ratio of base.
    delta = (rand() * 2 - 1) * policy.jitter * base
    return max(0, _round_half_up(base + delta))


async def with_provider_retry[T](
    fn: Callable[[], Awaitable[T]],
    *,
    policy: RetryPolicy,
    signal: AbortSignal,
    deadline: float,
    deps: RetryDeps | None = None,
) -> T:
    """Run `fn` and, on a retryable PROVIDER failure, retry with exponential backoff.

    Anything else (abort / non-retryable PROVIDER / schema-derived, etc.) is re-raised without being caught.
    If a wait would cross `deadline` (absolute time in ms = now() + timeout), give up without waiting.
    """
    d = deps if deps is not None else default_retry_deps
    attempt = 0
    while True:
        try:
            return await fn()
        except BaseException as err:
            # Do not retry on abort/timeout (if the shared signal is aborted, further attempts are pointless).
            if _is_abort_like(err, signal):
                raise
            retryable, retry_after_ms = retryable_provider_error(err)
            # Propagate immediately if not retryable or the limit is reached.
            if not retryable or attempt >= policy.max_retries:
                raise
            wait_ms = next_delay_ms(policy, attempt, retry_after_ms, d.random)
            # If waiting would only cross the deadline and get timed out right after, give up now without waiting.
            if d.now() + wait_ms >= deadline:
                raise
            await d.sleep(wait_ms, signal)
            attempt += 1

File: packages/llm/src/adapters/retry.ts
/**
 * Exponential backoff with jitter, limited to transient PROVIDER failures (429 / 5xx and the like — API errors
 * that the AI SDK judges as `isRetryable=true`). This single file centrally manages the retry policy, and the
 * SDK's built-in retries (default maxRetries=2) are pinned to 0 by the caller to avoid double retries.
 *
 * Design-level separation:
 * - ABORTED (caller cancellation / timeout overrun) propagates immediately and is not retried.
 * - INVALID_OUTPUT / CONFIG are out of scope (schema-derived failures are not fixed by retrying).
 * - PROVIDER with `isRetryable=false` (e.g. ollama's structured-output 400) is also out of scope. That is a
 *   structured-output incompatibility, which is the prompt JSON fallback's (auto) job. Retrying here would be a
 *   "fallback × backoff" double consumption, so it propagates immediately.
 * - The whole sequence never exceeds timeoutMs (deadline). If the next wait would not fit the remaining budget, stop.
 */
import type { RetryPolicy } from "../env.js";

/** Side effects that can be injected externally for determinism (tests swap wait / time / randomness). */
export interface RetryDeps {
  /** A wait that respects the AbortSignal. On abort, rejects with signal.reason. */
  sleep: (ms: number, signal: AbortSignal) => Promise<void>;
  now: () => number;
  /** Uniform random in [0,1) (for jitter). */
  random: () => number;
}

export const defaultRetryDeps: RetryDeps = {
  sleep: defaultSleep,
  now: () => Date.now(),
  random: () => Math.random(),
};

/** setTimeout-based wait that respects the AbortSignal. On abort, rejects with signal.reason. */
export function defaultSleep(ms: number, signal: AbortSignal): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    if (signal.aborted) {
      reject(signal.reason);
      return;
    }
    const onAbort = (): void => {
      clearTimeout(timer);
      reject(signal.reason);
    };
    const timer = setTimeout(() => {
      signal.removeEventListener("abort", onAbort);
      resolve();
    }, ms);
    // Do not let a backoff wait keep the process alive on shutdown (abort still clears the timer; optional
    // call because non-Node runtimes return a number here).
    (timer as { unref?: () => void }).unref?.();
    signal.addEventListener("abort", onAbort, { once: true });
  });
}

/**
 * Runs `fn` and, on a retryable PROVIDER failure, retries with exponential backoff.
 * Everything else (abort / non-retryable PROVIDER / schema-derived, etc.) is not caught and re-thrown as is.
 */
export async function withProviderRetry<T>(
  fn: () => Promise<T>,
  opts: {
    policy: RetryPolicy;
    signal: AbortSignal;
    /** Deadline for retry scheduling (absolute time in milliseconds). Given as now() + timeoutMs. */
    deadline: number;
    deps?: Partial<RetryDeps>;
  },
): Promise<T> {
  const { policy, signal, deadline } = opts;
  const deps: RetryDeps = { ...defaultRetryDeps, ...opts.deps };
  let attempt = 0;
  for (;;) {
    try {
      return await fn();
    } catch (err) {
      // Do not retry on abort / timeout (if the shared signal has expired, further attempts are pointless too).
      if (isAbortLike(err, signal)) throw err;
      const retryAfterMs = retryableProviderError(err);
      // Not retryable, or the limit is reached, so propagate immediately.
      if (retryAfterMs === undefined || attempt >= policy.maxRetries) throw err;
      const waitMs = nextDelayMs(policy, attempt, retryAfterMs, deps.random);
      // If waiting would only cross the deadline and be timeout-aborted right after, give up now without waiting.
      if (deps.now() + waitMs >= deadline) throw err;
      await deps.sleep(waitMs, signal);
      attempt += 1;
    }
  }
}

/** Whether the error stems from an abort / timeout (judged by name or signal expiry). */
function isAbortLike(err: unknown, signal: AbortSignal): boolean {
  if (signal.aborted) return true;
  const name = (err as { name?: unknown } | null | undefined)?.name;
  return name === "AbortError" || name === "TimeoutError";
}

/**
 * The next backoff wait (milliseconds). If the provider returned a Retry-After, respect it;
 * otherwise add ±jitter perturbation to base = initialDelayMs * factor^attempt.
 */
export function nextDelayMs(
  policy: RetryPolicy,
  attempt: number,
  retryAfterMs: number | null,
  random: () => number,
): number {
  if (retryAfterMs != null) return Math.max(0, Math.round(retryAfterMs));
  const base = policy.initialDelayMs * policy.backoffFactor ** attempt;
  // Map random() ∈ [0,1) to [-1,1) and perturb by a ratio of base.
  const delta = (random() * 2 - 1) * policy.jitter * base;
  return Math.max(0, Math.round(base + delta));
}

/**
 * Judges whether a PROVIDER failure is retryable and, if so, returns the Retry-After (milliseconds).
 * - undefined: not retryable (non-APICallError / isRetryable=false / schema-derived, etc.)
 * - null: retryable, no Retry-After (leave it to exponential backoff)
 * - number: retryable, Retry-After present (respect this wait)
 */
export function retryableProviderError(err: unknown): number | null | undefined {
  const api = findApiCallError(err);
  if (api == null || api.isRetryable !== true) return undefined;
  return retryAfterFromHeaders(api.responseHeaders);
}

/** Whether a PROVIDER failure is retryable (the boolean version used for the auto fallback branch decision). */
export function isRetryableProviderError(err: unknown): boolean {
  return retryableProviderError(err) !== undefined;
}

interface ApiCallErrorLike {
  isRetryable?: unknown;
  responseHeaders?: Record<string, string>;
  statusCode?: unknown;
  url?: unknown;
  cause?: unknown;
}

/**
 * Structurally searches for the AI SDK's APICallError (without adding a direct dependency on
 * `@ai-sdk/provider`; also follows cause). An APICallError always has `isRetryable` (boolean). To avoid
 * false positives with other boolean fields, it is required to also carry one of responseHeaders / statusCode / url.
 */
function findApiCallError(err: unknown, depth = 0): ApiCallErrorLike | undefined {
  if (err == null || typeof err !== "object" || depth > 4) return undefined;
  const e = err as ApiCallErrorLike;
  if (typeof e.isRetryable === "boolean" && ("responseHeaders" in e || "statusCode" in e || "url" in e)) {
    return e;
  }
  return findApiCallError(e.cause, depth + 1);
}

/** Interprets Retry-After-equivalent headers (retry-after-ms / retry-after). Follows the AI SDK's implementation. */
function retryAfterFromHeaders(headers: Record<string, string> | undefined): number | null {
  if (headers == null) return null;
  const retryAfterMs = headers["retry-after-ms"];
  if (retryAfterMs != null) {
    const ms = Number.parseFloat(retryAfterMs);
    if (Number.isFinite(ms) && ms >= 0) return ms;
  }
  const retryAfter = headers["retry-after"];
  if (retryAfter != null) {
    // Seconds (e.g. "2") or an HTTP date (e.g. "Wed, 21 Oct 2025 07:28:00 GMT").
    const seconds = Number.parseFloat(retryAfter);
    if (Number.isFinite(seconds) && seconds >= 0) return seconds * 1000;
    const dateMs = Date.parse(retryAfter) - Date.now();
    if (Number.isFinite(dateMs) && dateMs >= 0) return dateMs;
  }
  return null;
}
